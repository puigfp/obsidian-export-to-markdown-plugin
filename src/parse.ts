import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkGfm from "remark-gfm";
import remarkStringify from "remark-stringify";
import remarkWikiLink from "@portaljs/remark-wiki-link";
import remarkFrontmatter from "remark-frontmatter";
import { Root } from "remark-parse/lib";

const processor = unified()
	.use(remarkParse)
	.use(remarkGfm)
	.use(remarkWikiLink, {
		aliasDivider: "|", // by default, this extension uses [[Link:Display]] instead of [[Link|Display]] like Obsidian does
	})
	.use(remarkFrontmatter, ["yaml"]) // we don't do anything with the YAML frontmatter, but prevents parts of it from being parsed as Markdown titles because of the "---" delimiters
	.use(remarkStringify);

export function parseMarkdown(markdown: string): Root {
	const tree = processor.parse(markdown);
	return tree;
}

export function stringifyMarkdownTree(tree: Root): string {
	return processor.stringify(tree);
}
