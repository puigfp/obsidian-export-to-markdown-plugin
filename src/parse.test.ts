import { parseMarkdown, stringifyMarkdownTree } from "./parse";


describe("parse", () => {
  it("should parse markdown", async () => {
    const markdown = "# Hello, world! [[WikiLink Target|WikiLink Display Text]]";
    const tree = parseMarkdown(markdown);
    const stringified = stringifyMarkdownTree(tree);
    console.log(JSON.stringify(tree, null, 2));
    console.log(stringified);
    expect(tree).toMatchSnapshot();
  });
});
