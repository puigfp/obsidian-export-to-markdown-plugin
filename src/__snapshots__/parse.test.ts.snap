// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`parse should parse markdown 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 17,
              "line": 1,
              "offset": 16,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Hello, world! ",
        },
        {
          "data": {
            "alias": "WikiLink Display Text",
            "exists": false,
            "hChildren": [
              {
                "type": "text",
                "value": "WikiLink Display Text",
              },
            ],
            "hName": "a",
            "hProperties": {
              "className": "internal new",
              "href": "#/page/wikilink_target",
            },
            "permalink": "wikilink_target",
          },
          "position": {
            "end": {
              "column": 58,
              "line": 1,
              "offset": 57,
            },
            "start": {
              "column": 17,
              "line": 1,
              "offset": 16,
            },
          },
          "type": "wikiLink",
          "value": "WikiLink Target",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 58,
          "line": 1,
          "offset": 57,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
    },
  ],
  "position": {
    "end": {
      "column": 58,
      "line": 1,
      "offset": 57,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
